### ISSUE TITLE
SortedSet<T> probably needs re-implementing

### ISSUE THREAD

**Issue by metalgearsloth** [space-wizards/RobustToolbox#2488]
![image](https://user-images.githubusercontent.com/31366439/151960654-e2c981f8-cdce-40f5-b4f7-460d3dbf7154.png)

If I spawn 40 dummy pathfinding mobs it's no.1 and no.2 on allocs.

**Comment by Acruid** [comment]
`SortedSet<T>` is a red-black binary search tree, so the node allocations are expected. Are you sure pathfinding cannot use another data structure?

**Comment by metalgearsloth** [comment]
> `SortedSet<T>` is a red-black binary search tree, so the node allocations are expected. Are you sure pathfinding cannot use another data structure?

Does it need to allocate so much just to iterate the tree?

**Comment by metalgearsloth** [comment]
PJB also suggested potentially using a HashSet<T> but would need to profile to compare. Iterating is much more common than insertions / removals so that's probably the most important factor.

**Comment by Ygg01** [comment]
Here are operations on various sets.
![IMG_20220202_162549](https://user-images.githubusercontent.com/1146204/152183951-b2194632-c302-4588-86cc-2d43bdaa6cd1.jpg)

If this is an unsorted set of ints you can use FastSet. Or another structure.

**Comment by Acruid** [comment]
> > `SortedSet<T>` is a red-black binary search tree, so the node allocations are expected. Are you sure pathfinding cannot use another data structure?
> 
> Does it need to allocate so much just to iterate the tree?

It should not be allocating any nodes when iterating/traversing the tree, only modifying the tree. I can't find where SortedSet is used in pathfinding, but I do see it used in TransformComponent.Children?
![image](https://user-images.githubusercontent.com/1685952/152228023-67c51886-9c2b-4568-87d2-8de169e30b5c.png)

**Comment by metalgearsloth** [comment]
> I can't find where SortedSet is used in pathfinding, but I do see it used in TransformComponent.Children?

I just used the mobs to have something moving and it's just iterating the children allocating a stupid amount in entitylookup and pvs updates. Like almost as much as the rest of the game combined on the server.

When I swapped it to a hashset this effectively disappeared because it doesn't have the _stack = new Stack<T> that SortedSet<T> enumeration does.

I don't think we actually rely on the set being sorted anywhere in the codebase; deletion uses a ToArray for safety anyway and maploader doesn't make use of it.
